### PR TITLE
Added scraper for veroniquecloutier.com

### LINKED ISSUE #1252: Add scraper for veroniquecloutier.com
Please add support for scraping veroniquecloutier.com. Thanks in advance.

Recipe Example:
- https://veroniquecloutier.com/cuisine/boeuf-braise-a-loignon-et-puree-de-pommes-de-terre-au-bacon
- https://veroniquecloutier.com/cuisine/recette-roti-epaule-de-porc-facon-porchetta-de-dan-daniel-raiche-blake-mackay

### PR DESCRIPTION
**Pull request by shinyumh**
Resolves #1252

Added support for [veroniquecloutier.com](https://veroniquecloutier.com). Since this is a French recipe website, recipe schema is not available. Thus, I tried my best to identify patterns of where information is located. This is especially true for functions like `yields()`, `ingredients()`, and `instructions()`. I also chose not to implement `total_time()` since most of the recipes I looked at did not have this information.

Please let me know if there are any suggestions/feedback to make this better! Thank you in advance.

### CODE REVIEW

File: recipe_scrapers/veroniquecloutier.py [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+        ingredient_list = []
+
+        if start and end:
+            for sibling in start.find_all_next():
+                if sibling.string and "préparation" in sibling.string.lower():
+                    break
+                if sibling.name == "ul":
+                    ingredient_list.extend([li.text for li in sibling.find_all("li")])
+
+        return ingredient_list
```
Comment (jknndy) [inline recipe_scrapers/veroniquecloutier.py:80]:
```suggestion
    def ingredients(self):
        start = self.soup.find(
            string=lambda text: text and text.lower() == "ingrédients"
        )
        end = self.soup.find(string=lambda text: text and text.lower() == "préparation")

        if not start or not end:
            return []

        ingredient_list = []
        for sibling in start.find_all_next():
            if sibling.string and "préparation" in sibling.string.lower():
                break
            if sibling.name == "ul":
                ingredient_list.extend(li.text for li in sibling.find_all("li"))

        return ingredient_list
```
Added a check to return an empty list if either the start or end marker is not found  & removed the explicit list extension logic by directly using a generator for more Pythonic code.

File: recipe_scrapers/veroniquecloutier.py [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+        if start:
+            for sibling in start.find_all_next():
+                if sibling.name == "div":
+                    break
+                if sibling.name == "ol":
+                    instruction_list.extend([li.text for li in sibling.find_all("li")])
+                elif sibling.name == "p" and sibling.text[0].isdigit():
+                    instruction_list.append(sibling.text[3:])
+
+        return "\n".join(instruction_list)
```
Comment (jknndy) [inline recipe_scrapers/veroniquecloutier.py:98]:
```suggestion
    def instructions(self):
        start = self.soup.find(
            string=lambda text: text and text.lower() == "préparation"
        )

        if not start:
            return ""

        instruction_list = []
        for sibling in start.find_all_next():
            if sibling.name == "div":
                break
            if sibling.name == "ol":
                instruction_list.extend(li.text for li in sibling.find_all("li"))
            elif sibling.name == "p" and sibling.text[0].isdigit():
                instruction_list.append(sibling.text[3:])

        return "\n".join(instruction_list)
```
Added a check to return an empty string if the start marker isn’t found.

**Review by jknndy — CHANGES_REQUESTED** [review]
First off thanks for the contribution!
A few changes needed most of which are small improvements / ideas for optimization. Let me know what you think!

**Comment by jknndy** [comment]
Optional: if you wanted to try adding [ingredient_groups](https://github.com/hhursev/recipe-scrapers/blob/main/docs/in-depth-guide-ingredient-groups.md#in-depth-guide-ingredient-groups) and support for this site this [recipe ](https://veroniquecloutier.com/cuisine/boeuf-braise-a-loignon-et-puree-de-pommes-de-terre-au-bacon)contains it.

File: recipe_scrapers/__init__.py [hunk 3]
```diff
@@ -806,6 +808,8 @@
     VegRecipesOfIndia.host(): VegRecipesOfIndia,
     Vegetarbloggen.host(): Vegetarbloggen,
     Vegolosi.host(): Vegolosi,
+    VeroniqueCloutier.host(): VeroniqueCloutier,
```
Comment (jknndy) [inline recipe_scrapers/__init__.py:811]:
```suggestion
```

File: recipe_scrapers/veroniquecloutier.py [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+
+    def title(self):
+        return self.soup.find("h1", {"class": "title -main -page-title"}).get_text()
+
+    def category(self):
+        return (
+            self.soup.find("div", {"class": "categories"})
+            .findChild("span", {"class": "separator"})
+            .get_text()
+        )
```
[the remark this answers is not in the document]
Comment (shinyumh) [reply recipe_scrapers/veroniquecloutier.py]: I see; I removed the function!

File: recipe_scrapers/veroniquecloutier.py [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+        if potion_line:
+            words = potion_line.split()  # Split text into words
+            for word in words:
+                if word.isdigit():
+                    return str(word) + " portions"
+                elif word.lower() in french_numbers:
+                    return str(french_numbers.index(word.lower()) + 1) + " portions"
+                elif word.lower() in special_cases:
+                    return str(special_cases[word.lower()]) + " portions"
+        return None
```
[the remark this answers is not in the document]
Comment (shinyumh) [reply recipe_scrapers/veroniquecloutier.py]: Thank you so much for the detailed review - they make sense and definitely help with the readability of the code. I changed my test case to reflect this change as well.

File: recipe_scrapers/veroniquecloutier.py (hunk 1 above)
Comment (shinyumh) [reply recipe_scrapers/veroniquecloutier.py]: Thank you! I integrated the changes you made into the code.

**Comment by shinyumh** [comment]
> Optional: if you wanted to try adding [ingredient_groups](https://github.com/hhursev/recipe-scrapers/blob/main/docs/in-depth-guide-ingredient-groups.md#in-depth-guide-ingredient-groups) support for this site this [recipe ](https://veroniquecloutier.com/cuisine/boeuf-braise-a-loignon-et-puree-de-pommes-de-terre-au-bacon)contains it.

Thank you so much for reviewing my PR! I made the recommended changes and also added support for `ingredient_groups()` by implementing custom grouping logic. Let me know if there is anything else I should add/change!

**Comment by jknndy** [comment]
Looks great! One last thing that I forgot to mention last time, our test suite enforces a specific order of JSON keys in the test files for ease of review. There is a helper script built in that can handle reordering the keys into the correct order that can be run with 
```
python .\scripts\reorder_json_keys.py
```

**Comment by shinyumh** [comment]
Thanks for the catch! I ran the script to reorder JSON keys in the test file. Really appreciate you taking the time to review my PR!

**Review by jknndy — APPROVED** [review]

File: recipe_scrapers/veroniquecloutier.py [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+
+    def ingredients(self):
+        start = self.soup.find(
+            string=lambda text: text and text.lower() == "ingrédients"
+        )
+        end = self.soup.find(string=lambda text: text and text.lower() == "préparation")
+
+        if not start or not end:
+            return []
```
Comment (jayaddison) [inline recipe_scrapers/veroniquecloutier.py:72]: Looks pretty great - thank you @shinyumh (and also @jknndy for reviewing).

One question for you both: do we need these `if not  start ...:` conditions?

Explanation: if something doesn't contain either any ingredients, or doesn't contain any instructions, I'd generally be skeptical of whether it's a recipe -- and so I ideally perhaps we could remove this (but I'd like to check before suggesting that).

Comment (shinyumh) [reply recipe_scrapers/veroniquecloutier.py:72]: That is a good point. I just reviewed ~10 recipes on the website and it seemed like all of them have "Ingrédients" and "Préparation" section headers. I am open to removing these conditions if @jknndy also agrees.

Comment (jknndy) [reply recipe_scrapers/veroniquecloutier.py:72]: Agreed! Thanks @jayaddison

Comment (shinyumh) [reply recipe_scrapers/veroniquecloutier.py]: Fixed, thank you both!

**Review by hhursev — APPROVED** [review]

**Pull request merged**